compiler/rustc_resolve/src/access_levels.rs

```diff
@@ -96,9 +96,15 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
         parent_id: LocalDefId,
         tag: AccessLevel,
     ) {
+        let module_id = self
+            .r
+            .get_nearest_non_block_module(def_id.to_def_id())
+            .nearest_parent_mod()
+            .expect_local();
+        if nominal_vis == Visibility::Restricted(module_id) {
```

Comment:
```suggestion
        if nominal_vis == Visibility::Restricted(module_id) || self.r.visibilities[&parent_id] == Visibilities::Restricted(module_id) {
```
Both `nominal_vis` and `parent_id` set an upper bound on effective visibility, so this condition can be extended like this.